tests/unit/lib/transport/test_mqtt.py

```diff
@@ -0,0 +1,821 @@
+        with self.assertRaises(ValueError):
+            Mqtt(self.url, self.port, credentials, tls_conf, qos_details, self.client_id,
+                 self.client_clean_session, self.user_data, self.protocol, self.transport,
+                 self.keep_alive, self.enable_authentication, self.connection_disconnect_timeout)
+
+    def test_connect_soc_for_empty_username(self):
+        """
+        Test case to test validation for empty username.
+        :return: None
+        """
+
+        # Encapsulate the authentication details
+        credentials = Identity(self.root_ca_cert, "", self.mqtt_password,
+                               self.client_cert_file, self.client_key_file)
+
+        # Encapsulate TLS parameters
+        tls_conf = TLSConf(self.cert_required, self.tls_version, self.cipher)
+
+        # Encapsulate QoS related parameters
+        qos_details = QoSDetails(self.inflight, self.queue_size, self.retry)
+
+        # Checking whether implementation raising the ValueError for invalid username
+        with self.assertRaises(ValueError):
+            Mqtt(self.url, self.port, credentials, tls_conf, qos_details, self.client_id,
+                 self.client_clean_session, self.user_data, self.protocol, self.transport,
+                 self.keep_alive, self.enable_authentication, self.connection_disconnect_timeout)
+
+    def test_connect_soc_for_empty_password(self):
+        """
+        Test case to test validation for empty password.
+        :return: None
+        """
+
+        # Encapsulate the authentication details
+        credentials = Identity(self.root_ca_cert, self.mqtt_username, "",
+                               self.client_cert_file, self.client_key_file)
+
+        # Encapsulate TLS parameters
+        tls_conf = TLSConf(self.cert_required, self.tls_version, self.cipher)
+
+        # Encapsulate QoS related parameters
+        qos_details = QoSDetails(self.inflight, self.queue_size, self.retry)
+
+        # Checking whether implementation raising the ValueError for empty password
+        with self.assertRaises(ValueError):
+            Mqtt(self.url, self.port, credentials, tls_conf, qos_details, self.client_id,
+                 self.client_clean_session, self.user_data, self.protocol, self.transport,
+                 self.keep_alive, self.enable_authentication, self.connection_disconnect_timeout)
+
+    def test_connect_soc_connection_setup(self):
+        """
+        Test case to test connection setup with connect_soc.
+        :return: None
+        """
+        global connect_rc
+
+        # Setting connection accepted flag
+        connect_rc = 0
+
+        # Mocked the connect and loop_start method of Paho library
+        Client.connect = mocked_connect
+        Client.loop_start = mocked_loop_start
+
+        # Encapsulate the authentication details
+        credentials = Identity(self.root_ca_cert, self.mqtt_username, self.mqtt_password,
```

Comment:
change credentials to identity.
I see that identity, tls_conf and qos_details are use in all test cases.  So, create these in setup and have appropriate teardown
For invalid root_ca, username, password and related testcases, create within than testcase.